File: backend/src/providers/cartProvider.ts
import { prisma } from '../lib/prisma.js';

export const cartProvider = {
  async findByUserId(userId: bigint) {
    return prisma.carts.findUnique({
      where: { user_id: userId, deleted: false },
      include: {
        cart_items: {
          where: { deleted: false },
          include: {
            menu_items: {
              include: {
                menu_item_images: {
                  where: { is_primary: true },
                  take: 1,
                },
              },
            },
          },
        },
      },
    });
  },

  async findBySessionId(sessionId: string) {
    return prisma.carts.findFirst({
      where: { session_id: sessionId, user_id: null, deleted: false },
      include: {
        cart_items: {
          where: { deleted: false },
          include: {
            menu_items: {
              include: {
                menu_item_images: {
                  where: { is_primary: true },
                  take: 1,
                },
              },
            },
          },
        },
      },
    });
  },

  async create(data: { user_id?: bigint; session_id?: string }) {
    return prisma.carts.create({
      data: {
        ...data,
      },
    });
  },

  async findItem(cartId: bigint, menuItemId: bigint, modifications: string[]) {
    // Note: modifications is an array, comparing arrays in SQL can be tricky.
    // Prisma's equals for string[] should work if order matches.
    return prisma.cart_items.findFirst({
      where: {
        cart_id: cartId,
        menu_item_id: menuItemId,
        modifications: { equals: modifications },
        deleted: false,
      },
    });
  },

  async addItem(data: {
    cart_id: bigint;
    menu_item_id: bigint;
    quantity: number;
    modifications: string[];
  }) {
    return prisma.cart_items.create({
      data,
    });
  },

  async updateItem(
    itemId: bigint,
    data: { quantity?: number; modifications?: string[] },
  ) {
    return prisma.cart_items.update({
      where: { id: itemId },
      data,
    });
  },

  async removeItem(itemId: bigint) {
    return prisma.cart_items.update({
      where: { id: itemId },
      data: { deleted: true },
    });
  },

  async deleteItem(itemId: bigint) {
    return prisma.cart_items.delete({
      where: { id: itemId },
    });
  },

  async mergeCartItems(fromCartId: bigint, toCartId: bigint) {
    // This is more complex, usually done in service
    // But we can have a provider method to update all items' cart_id
    return prisma.cart_items.updateMany({
      where: { cart_id: fromCartId, deleted: false },
      data: { cart_id: toCartId },
    });
  },

  async deleteCart(cartId: bigint) {
    return prisma.carts.update({
      where: { id: cartId },
      data: { deleted: true },
    });
  },

  async clearItemsByCartId(cartId: bigint) {
    return prisma.cart_items.updateMany({
      where: { cart_id: cartId, deleted: false },
      data: { deleted: true },
    });
  },
};

File: backend/app/providers/cart_provider.py
from sqlalchemy.ext.asyncio import AsyncSession
from sqlalchemy import select, update, delete
from sqlalchemy.orm import selectinload
from typing import Optional, List
from app.models.cart import Cart
from app.models.cart_item import CartItem
from app.models.menu_item import MenuItem
from app.models.menu_item_image import MenuItemImage

class CartProvider:
    async def find_by_user_id(self, db: AsyncSession, user_id: int) -> Optional[Cart]:
        result = await db.execute(
            select(Cart)
            .options(selectinload(Cart.cart_items).selectinload(CartItem.menu_item).selectinload(MenuItem.menu_item_images))
            .where(Cart.user_id == user_id, Cart.deleted == False)
        )
        return result.scalar_one_or_none()

    async def find_by_session_id(self, db: AsyncSession, session_id: str) -> Optional[Cart]:
        result = await db.execute(
            select(Cart)
            .options(selectinload(Cart.cart_items).selectinload(CartItem.menu_item).selectinload(MenuItem.menu_item_images))
            .where(Cart.session_id == session_id, Cart.user_id.is_(None), Cart.deleted == False)
        )
        return result.scalar_one_or_none()

    async def create(self, db: AsyncSession, user_id: Optional[int] = None, session_id: Optional[str] = None) -> Cart:
        cart = Cart(user_id=user_id, session_id=session_id)
        db.add(cart)
        await db.commit()
        await db.refresh(cart)
        return cart

    async def find_item(self, db: AsyncSession, cart_id: int, menu_item_id: int, modifications: List[str]) -> Optional[CartItem]:
        result = await db.execute(
            select(CartItem).where(
                CartItem.cart_id == cart_id,
                CartItem.menu_item_id == menu_item_id,
                CartItem.modifications == modifications,
                CartItem.deleted == False,
            )
        )
        return result.scalar_one_or_none()

    async def add_item(self, db: AsyncSession, cart_id: int, menu_item_id: int, quantity: int, modifications: List[str]) -> CartItem:
        item = CartItem(cart_id=cart_id, menu_item_id=menu_item_id, quantity=quantity, modifications=modifications)
        db.add(item)
        await db.commit()
        await db.refresh(item)
        return item

    async def update_item(self, db: AsyncSession, item_id: int, **data) -> Optional[CartItem]:
        await db.execute(update(CartItem).where(CartItem.id == item_id).values(**data))
        await db.commit()
        result = await db.execute(select(CartItem).where(CartItem.id == item_id))
        return result.scalar_one_or_none()

    async def remove_item(self, db: AsyncSession, item_id: int) -> None:
        await db.execute(update(CartItem).where(CartItem.id == item_id).values(deleted=True))
        await db.commit()

    async def delete_item(self, db: AsyncSession, item_id: int) -> None:
        await db.execute(delete(CartItem).where(CartItem.id == item_id))
        await db.commit()

    async def delete_cart(self, db: AsyncSession, cart_id: int) -> None:
        await db.execute(update(Cart).where(Cart.id == cart_id).values(deleted=True))
        await db.commit()

    async def clear_items_by_cart_id(self, db: AsyncSession, cart_id: int) -> None:
        await db.execute(update(CartItem).where(CartItem.cart_id == cart_id, CartItem.deleted == False).values(deleted=True))
        await db.commit()

cart_provider = CartProvider()
